File: web/lib/language.ts
/** The bucket an item with no language reading falls in. `null` is unmeasured, not a language,
 *  so it gets its own key rather than being folded into en or dropped out of the set. */
export const NO_LANG = "none"

/** What a language code is called on the board. A language names itself in its own script where
 *  it has one — a Chinese speaker scanning this row should not have to read `zh`. Anything not
 *  listed carries through as its own code rather than being renamed to a guess. */
export const LANG_NAMES: Record<string, string> = {
  en: "english",
  zh: "中文",
  [NO_LANG]: "unread",
}

export type LangTab = { key: string; label: string; count: number }

/**
 * The language tabs for any set of things that carry a language.
 *
 * Shared by the channel directory, whose language is hand-authored per channel in
 * `config/channels.json`, and by the recent feed, whose language is read per video by
 * `pipeline/language.py`. They are different questions on different data, but the vocabulary and
 * the ordering are the same, and a second copy is how the two rows start disagreeing about what
 * `zh` is called.
 *
 * Callers pass the already-filtered set. The counts therefore run over what the viewer can
 * actually reach, so a tab's number is always what clicking it produces.
 */
export function langTabsFor<T extends { lang?: string | null }>(items: T[]): LangTab[] {
  const counts = new Map<string, number>()
  for (const it of items) {
    const k = it.lang ?? NO_LANG
    counts.set(k, (counts.get(k) ?? 0) + 1)
  }
  // Biggest language first; the unmeasured bucket sinks to the end whatever its size, because it
  // is a state and the tabs before it are a roster.
  const langs = [...counts.entries()]
    .sort((a, b) => {
      if (a[0] === NO_LANG) return 1
      if (b[0] === NO_LANG) return -1
      return b[1] - a[1] || a[0].localeCompare(b[0])
    })
    .map(([key, count]) => ({ key, label: LANG_NAMES[key] ?? key, count }))
  return [{ key: "all", label: "all", count: items.length }, ...langs]
}

File: pipeline/language.py
"""What language a video is in, and how confidently we know it.

Two signals, and the tier says which one answered. `defaultAudioLanguage` is the uploader's own
declaration, carried in the videos.list snippet the sweep already pays for — Oracle. The CJK share
of the title is a rule this project applies — Derived, and it ships the rule that produced it.
Neither firing is `unread`: its own state, never folded into en.

This is not the channel's `lang` in config/channels.json, which is hand-authored, one value per
channel, and is what /channels filters on. A channel that posts in both scripts has one value
there and a correct one per video here.

Nothing here reads a transcript or a description. Measured over the corpus on 2026-07-31, a title
check misreads 6 videos out of 11,859, and extraction is behind step 13's hard gate.
"""
from __future__ import annotations

import re

#: The bucket a video with no readable language falls in. Not a language.
NO_LANG = "none"

# CJK Unified Ideographs, Extension A, and the compatibility block. Kana and Hangul are
# deliberately absent: this board tracks an English and a Chinese scene, and a Japanese title
# should read as its own code from a declaration rather than be folded into zh by script alone.
_CJK = re.compile(r"[一-鿿㐀-䶿豈-﫿]")


def _primary_subtag(tag: str) -> str:
    """`zh-Hant` and `ZH-hans` are both zh. A tag this project has no opinion on carries through."""
    return tag.strip().lower().split("-")[0]


def cjk_share(title: str) -> float:
    """CJK characters as a share of the title's alphanumerics. 0.0 when there is nothing to count.

    The denominator is alphanumerics rather than the whole string so that punctuation, emoji, and
    the spaces around a Latin product name cannot dilute a Chinese sentence below the threshold.
    """
    letters = [c for c in title if c.isalnum()]
    if not letters:
        return 0.0
    return sum(1 for c in letters if _CJK.match(c)) / len(letters)


def detect(title: str | None, default_audio_language: str | None, threshold: float,
           description: str | None = None) -> tuple[str, str]:
    """Returns (lang, tier). tier is one of "oracle", "derived", "unread".

    The description breaks a tie, and only a tie. A title carrying some CJK but not enough to
    clear the threshold is the one case where the title genuinely cannot answer: a Chinese
    sentence padded out by a long Latin product name and a run of hashtags reads as English on
    character share alone. `Claude Code 一键切换到 DeepSeek (CC Switch) #Shorts #claudecode` is
    0.09 of a threshold of 0.10, and its description is 0.44.

    It is deliberately not consulted for an all-Latin title. Chinese creators put the same
    Chinese membership boilerplate under every upload, including their genuinely English ones, so
    a description-first rule would relabel a whole channel on the strength of a footer. Zero
    all-Latin titles in the corpus currently have a Chinese description, and this keeps it that
    way by construction rather than by luck.
    """
    if default_audio_language:
        return _primary_subtag(default_audio_language), "oracle"
    if not title:
        return NO_LANG, "unread"
    if not any(c.isalnum() for c in title):
        return NO_LANG, "unread"
    share = cjk_share(title)
    if share >= threshold:
        return "zh", "derived"
    if share > 0 and description and cjk_share(description) >= threshold:
        return "zh", "derived"
    return "en", "derived"
